README.md
# This is the original readme but much of it is out of date with the fork, ima redact it sometime soon
# Check out WhyFork.md

# Description
An osu! Bongo Cat overlay with smooth paw movement and simple skinning ability, written in C++. Originally created by [HamishDuncanson](https://github.com/HamishDuncanson).

You can find how to configure the application in our [wiki](https://github.com/kuroni/bongocat-osu/wiki/Settings).

Download the program [here](https://github.com/kuroni/bongocat-osu/releases).

Hugs and kisses to [CSaratakij](https://github.com/CSaratakij) for creating the Linux port for this project!

Any suggestion and/or collaboration, especially that relating to sprites, is welcomed! Thank you!

[Original post](https://www.reddit.com/r/osugame/comments/9hrkte/i_know_bongo_cat_is_getting_old_but_heres_a_nicer/) by [Kuvster](https://github.com/Kuvster).

## Further information
In order to play with fullscreen on Windows 10, run both osu! and this application in Windows 7 compability mode.

Press Ctrl + R to reload configuration and images (will only reload configurations when the window is focused).

Supported operating system:
* Windows
* Linux (tested with Arch Linux with WINE Staging 5). Note: You **must** use WINE Staging, because for whatever reason on stable WINE bongocat-osu doesn't register keyboard input from other windows.

_Notice_: If you're using WINE on Linux, make sure that osu! and this application run in the same `WINEPREFIX`.

## For developers
This project uses [SFML](https://www.sfml-dev.org/index.php) and [JsonCpp](https://github.com/open-source-parsers/jsoncpp). JsonCpp libraries are directly included in the source using the provided `amalgamation.py` from the developers.

### Libraries and dependency

#### Windows and MinGW
To build the source, download the SFML libraries [here](https://www.sfml-dev.org/index.php), copy `Makefile.windows` to `Makefile`, then replace *`<SFML-folder>`* in `Makefile` with the desired folder.

#### Linux
You need to have these dependencies installed. Check with your package manager for the exact name of these dependencies on your distro:
- g++
- libxdo
- sdl2
- sfml
- x11
- xrandr

Then, copy `Makefile.linux` to `Makefile`.

### Building and testing
To build, run this command from the base directory:

```sh
make
```

To test the program, run this from the base directory:

```sh
make test
```

Alternatively, you can copy the newly-compiled `bin/bongo.exe` or `bin/bongo` into the base directory and execute it.

If you have troubles compiling, it can be due to version mismatch between your compiler and SFML. See [#43](https://github.com/kuroni/bongocat-osu/issues/43) for more information.



changelog
# 2.0.0
- avatar is now on a separate layer from the desk, and on top of the arm, so the arm looks much better
- toggle debug view with CTRL-D
- removed distinction between mouse and tablet (achievable by skins)
- prepared file structure for skins
- filename changes
    - `cat.png` -> `avatar.png`
    - `mousebg.png` -> `background.png`
- config
    - removed entries relating to original hand drawing
    - `stretchyArm` key is no longer needed as there is no alternative

# 1.0.0
- stretchy arm slapped on top of original app

---
---
---

# backlog
- avatar semi transparent in debug\
- some images could be loaded with transparent as default
- support many monitors
- auto resize window to match the desk

src/ctb.cpp
#include "header.hpp"

namespace ctb {
Json::Value left_key_value, right_key_value, dash_key_value;
sf::Sprite bg, mid, left, right, dash, up;

int key_state = 0;
bool left_key_state = false;
bool right_key_state = false;
double timer_left_key = -1;
double timer_right_key = -1;

bool init() {
    // getting configs
    Json::Value ctb = data::cfg["catch"];

    bool chk[256];
    std::fill(chk, chk + 256, false);
    left_key_value = ctb["left"];
    for (Json::Value &v : left_key_value) {
        chk[v.asInt()] = true;
    }
    right_key_value = ctb["right"];
    for (Json::Value &v : right_key_value) {
        if (chk[v.asInt()]) {
            data::error_msg("Overlapping osu!catch keybinds", "Error reading configs");
            return false;
        }
    }
    dash_key_value = ctb["dash"];

    // importing sprites
    bg.setTexture(data::load_texture("img/catch/bg.png"));
    mid.setTexture(data::load_texture("img/catch/mid.png"));
    left.setTexture(data::load_texture("img/catch/left.png"));
    right.setTexture(data::load_texture("img/catch/right.png"));
    dash.setTexture(data::load_texture("img/catch/dash.png"));
    up.setTexture(data::load_texture("img/catch/up.png"));

    return true;
}

void draw() {
    window.draw(bg);

    // drawing left-right keypresses
    bool left_key = false;
    for (Json::Value &v : left_key_value) {
        if (input::is_pressed(v.asInt())) {
            left_key = true;
            break;
        }
    }
    if (left_key) {
        if (!left_key_state) {
            key_state = 1;
            left_key_state = true;
        }
    } else {
        left_key_state = false;
    }

    bool right_key = false;
    for (Json::Value &v : right_key_value) {
        if (input::is_pressed(v.asInt())) {
            right_key = true;
            break;
        }
    }
    if (right_key) {
        if (!right_key_state) {
            key_state = 2;
            right_key_state = true;
        }
    } else {
        right_key_state = false;
    }

    if (!left_key_state && !right_key_state) {
        key_state = 0;
        window.draw(mid);
    }
    if (key_state == 1) {
        if ((clock() - timer_right_key) / CLOCKS_PER_SEC > BONGO_KEYPRESS_THRESHOLD) {
            window.draw(left);
            timer_left_key = clock();
        } else {
            window.draw(mid);
        }
    } else if (key_state == 2) {
        if ((clock() - timer_left_key) / CLOCKS_PER_SEC > BONGO_KEYPRESS_THRESHOLD) {
            window.draw(right);
            timer_right_key = clock();
        } else {
            window.draw(mid);
        }
    }

    bool is_dash = false;
    for (Json::Value &v : dash_key_value) {
        if (input::is_pressed(v.asInt())) {
            window.draw(dash);
            is_dash = true;
            break;
        }
    }
    if (!is_dash) {
        window.draw(up);
    }
}
}; // namespace ctb


src/osu.cpp
#include "header.hpp"

namespace osu {
Json::Value left_key_value, right_key_value, smoke_key_value, wave_key_value;
int offset_x, offset_y;
int paw_r, paw_g, paw_b, paw_a;
int paw_edge_r, paw_edge_g, paw_edge_b, paw_edge_a;
double scale;
bool is_left_handed, is_enable_toggle_smoke;
sf::Sprite desk, cat, up, left, right, device, smoke, wave, arm;
bool debug;
sf::CircleShape anchorMark, handMark;
sf::RectangleShape catsMouseMark;

int key_state = 0;

bool left_key_state = false;
bool right_key_state = false;
bool wave_key_state = false;
bool previous_smoke_key_state = false;
bool current_smoke_key_state = false;
bool is_toggle_smoke = false;
bool stretchy_arm = false;

double timer_left_key = -1;
double timer_right_key = -1;
double timer_wave_key = -1;
const double PI = 3.141592653589793238462643383279502884;
const double toDeg = 180.0 / PI;
double catsMouseAngle;
// std::string skin = "default";
std::string skin = data::skin;
std::string skinpath = "skins/" + skin + "/";

void toggle_debug() {
    debug = !debug;
}

bool init() {
    // getting configs
    Json::Value osu = data::cfg;

    stretchy_arm = osu["stretchyArm"].asBool();
    debug = osu["debug"].asBool();
    if (stretchy_arm) {
        anchorMark.setFillColor(sf::Color::Blue);
        anchorMark.setRadius(5);

        handMark.setFillColor(sf::Color::Red);
        handMark.setRadius(5);

        catsMouseMark.setFillColor(sf::Color::Transparent);
        catsMouseMark.setOutlineColor(sf::Color::Yellow);
        catsMouseMark.setOutlineThickness(2);

        double x, y;
        x = osu["armAnchorOffset"][0].asDouble();
        y = osu["armAnchorOffset"][1].asDouble();
        arm.setOrigin(x, y);

        x = osu["anchor"][0].asDouble();
        y = osu["anchor"][1].asDouble();
        anchorMark.setPosition(x, y);
        arm.setPosition(x, y);

        x = osu["catsMouse"][0].asDouble();
        y = osu["catsMouse"][1].asDouble();
        catsMouseMark.setPosition(x, y);
        x = osu["catsMouse"][2].asDouble();
        y = osu["catsMouse"][3].asDouble();
        catsMouseMark.setSize(sf::Vector2f(x, y));
        catsMouseAngle = osu["catsMouse"][4].asDouble();
        catsMouseMark.setRotation(catsMouseAngle);
    }
    is_enable_toggle_smoke = osu["toggleSmoke"].asBool();

    paw_r = osu["paw"][0].asInt();
    paw_g = osu["paw"][1].asInt();
    paw_b = osu["paw"][2].asInt();
    paw_a = osu["paw"].size() == 3 ? 255 : osu["paw"][3].asInt();

    paw_edge_r = osu["pawEdge"][0].asInt();
    paw_edge_g = osu["pawEdge"][1].asInt();
    paw_edge_b = osu["pawEdge"][2].asInt();
    paw_edge_a = osu["pawEdge"].size() == 3 ? 255 : osu["pawEdge"][3].asInt();

    bool chk[256];
    std::fill(chk, chk + 256, false);
    left_key_value = osu["key1"];
    for (Json::Value &v : left_key_value) {
        chk[v.asInt()] = true;
    }
    right_key_value = osu["key2"];
    for (Json::Value &v : right_key_value) {
        if (chk[v.asInt()]) {
            data::error_msg("Overlapping osu! keybinds", "Error reading configs");
            return false;
        }
    }
    wave_key_value = osu["wave"];
    for (Json::Value &v : wave_key_value) {
        if (chk[v.asInt()]) {
            data::error_msg("Overlapping osu! keybinds", "Error reading configs");
            return false;
        }
    }
    smoke_key_value = osu["smoke"];

    is_left_handed = data::cfg["decoration"]["leftHanded"].asBool();

    offset_x = (data::cfg["decoration"]["offsetX"])[0].asInt();
    offset_y = (data::cfg["decoration"]["offsetY"])[0].asInt();
    scale = (data::cfg["decoration"]["scalar"])[0].asDouble();

    // importing sprites
    skin = data::skin;
    skinpath = "skins/" + skin + "/";
    up.setTexture(data::load_texture(skinpath + "up.png"));
    left.setTexture(data::load_texture(skinpath + "left.png"));
    right.setTexture(data::load_texture(skinpath + "right.png"));
    wave.setTexture(data::load_texture(skinpath + "wave.png"));
    cat.setTexture(data::load_texture(skinpath + "avatar.png"));
    desk.setTexture(data::load_texture(skinpath + "background.png"));
    device.setTexture(data::load_texture(skinpath + "mouse.png"), true);
    smoke.setTexture(data::load_texture(skinpath + "smoke.png"));
    arm.setTexture(data::load_texture(skinpath + "arm.png"));
    device.setScale(scale, scale);

    if (stretchy_arm) {
        auto b = device.getLocalBounds();
        device.setOrigin(b.width / 2, b.height / 2);
    }
    else {
        device.setOrigin(0, 0);
    }

    return true;
}

void draw_arm_fill(std::vector<double>& pss, std::vector<double>& pss2) {
    sf::VertexArray fill(sf::TriangleStrip, 26);
    for (int i = 0; i < 26; i += 2) {
        fill[i].position = sf::Vector2f(pss2[i], pss2[i + 1]);
        fill[i + 1].position = sf::Vector2f(pss2[52 - i - 2], pss2[52 - i - 1]);
        fill[i].color = sf::Color(paw_r, paw_g, paw_b, paw_a);
        fill[i + 1].color = sf::Color(paw_r, paw_g, paw_b, paw_a);
    }
    window.draw(fill);
}

void draw_arm_outline(std::vector<double>& pss, std::vector<double>& pss2) {
    // drawing first arm arc (solid black, rough edges)
    int shad = paw_edge_a / 3;
    sf::VertexArray edge(sf::TriangleStrip, 52);
    double width = 7;
    sf::CircleShape circ(width / 2);
    circ.setFillColor(sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, shad));
    circ.setPosition(pss2[0] - width / 2, pss2[1] - width / 2);
    window.draw(circ);
    for (int i = 0; i < 50; i += 2) {
        double vec0 = pss2[i] - pss2[i + 2];
        double vec1 = pss2[i + 1] - pss2[i + 3];
        double dist = hypot(vec0, vec1);
        edge[i].position = sf::Vector2f(pss2[i] + vec1 / dist * width / 2, pss2[i + 1] - vec0 / dist * width / 2);
        edge[i + 1].position = sf::Vector2f(pss2[i] - vec1 / dist * width / 2, pss2[i + 1] + vec0 / dist * width / 2);
        edge[i].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, shad);
        edge[i + 1].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, shad);
        width -= 0.08;
    }
    double vec0 = pss2[50] - pss2[48];
    double vec1 = pss2[51] - pss2[49];
    double dist = hypot(vec0, vec1);
    edge[51].position = sf::Vector2f(pss2[50] + vec1 / dist * width / 2, pss2[51] - vec0 / dist * width / 2);
    edge[50].position = sf::Vector2f(pss2[50] - vec1 / dist * width / 2, pss2[51] + vec0 / dist * width / 2);
    edge[50].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, shad);
    edge[51].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, shad);
    window.draw(edge);
    circ.setRadius(width / 2);
    circ.setPosition(pss2[50] - width / 2, pss2[51] - width / 2);
    window.draw(circ);

    // drawing second arm arc (shadow)
    sf::VertexArray edge2(sf::TriangleStrip, 52);
    width = 6;
    sf::CircleShape circ2(width / 2);
    circ2.setFillColor(sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, paw_edge_a));
    circ2.setPosition(pss2[0] - width / 2, pss2[1] - width / 2);
    window.draw(circ2);
    for (int i = 0; i < 50; i += 2) {
        vec0 = pss2[i] - pss2[i + 2];
        vec1 = pss2[i + 1] - pss2[i + 3];
        double dist = hypot(vec0, vec1);
        edge2[i].position = sf::Vector2f(pss2[i] + vec1 / dist * width / 2, pss2[i + 1] - vec0 / dist * width / 2);
        edge2[i + 1].position = sf::Vector2f(pss2[i] - vec1 / dist * width / 2, pss2[i + 1] + vec0 / dist * width / 2);
        edge2[i].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, paw_edge_a);
        edge2[i + 1].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, paw_edge_a);
        width -= 0.08;
    }
    vec0 = pss2[50] - pss2[48];
    vec1 = pss2[51] - pss2[49];
    dist = hypot(vec0, vec1);
    edge2[51].position = sf::Vector2f(pss2[50] + vec1 / dist * width / 2, pss2[51] - vec0 / dist * width / 2);
    edge2[50].position = sf::Vector2f(pss2[50] - vec1 / dist * width / 2, pss2[51] + vec0 / dist * width / 2);
    edge2[50].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, paw_edge_a);
    edge2[51].color = sf::Color(paw_edge_r, paw_edge_g, paw_edge_b, paw_edge_a);
    window.draw(edge2);
    circ2.setRadius(width / 2);
    circ2.setPosition(pss2[50] - width / 2, pss2[51] - width / 2);
    window.draw(circ2);
}

void draw_stretchy_arm() {
    auto [x, y] = input::where_mouse();

    sf::Transform t;
    t.translate(catsMouseMark.getPosition());
    t.rotate(catsMouseAngle, 0, 0);
    t.scale(catsMouseMark.getSize());
    sf::Vector2f handPos = t.transformPoint(sf::Vector2f(x, y));

    handMark.setPosition(handPos);
    sf::Vector2f displacement = handPos - anchorMark.getPosition();

    double dist = hypot(displacement.x, displacement.y);
    double scale = dist / arm.getLocalBounds().height;
    arm.setScale(1, scale);
    double alpha = asin(-displacement.x / dist);
    double deg = alpha * toDeg;
    arm.setRotation(deg);

    device.setPosition(handPos);
    window.draw(device);

    window.draw(arm);
}

void draw() {
    window.draw(desk);

    draw_stretchy_arm();

    window.draw(cat);
    if (debug && stretchy_arm) {
        window.draw(handMark);
        window.draw(anchorMark);
        window.draw(catsMouseMark);
    }
    // drawing keypresses
    bool left_key = false;

    for (Json::Value &v : left_key_value) {
        if (input::is_pressed(v.asInt())) {
            left_key = true;
            break;
        }
    }

    if (left_key) {
        if (!left_key_state) {
            key_state = 1;
            left_key_state = true;
        }
    } else {
        left_key_state = false;
    }

    bool right_key = false;

    for (Json::Value &v : right_key_value) {
        if (input::is_pressed(v.asInt())) {
            right_key = true;
            break;
        }
    }

    if (right_key) {
        if (!right_key_state) {
            key_state = 2;
            right_key_state = true;
        }
    } else {
        right_key_state = false;
    }

    bool wave_key = false;

    for (Json::Value &v : wave_key_value) {
        if (input::is_pressed(v.asInt())) {
            wave_key = true;
            break;
        }
    }

    if (wave_key) {
        if (!wave_key_state) {
            key_state = 3;
            wave_key_state = true;
        }
    } else {
        wave_key_state = false;
    }

    if (!left_key_state && !right_key_state && !wave_key_state) {
        key_state = 0;
        window.draw(up);
    }

    if (key_state == 1) {
        if ((clock() - std::max(timer_right_key, timer_wave_key)) / CLOCKS_PER_SEC > BONGO_KEYPRESS_THRESHOLD) {
            if (!is_left_handed) {
                window.draw(left);
            } else {
                window.draw(right);
            }
            timer_left_key = clock();
        } else {
            window.draw(up);
        }
    } else if (key_state == 2) {
        if ((clock() - std::max(timer_left_key, timer_wave_key)) / CLOCKS_PER_SEC > BONGO_KEYPRESS_THRESHOLD) {
            if (!is_left_handed) {
                window.draw(right);
            } else {
                window.draw(left);
            }
            timer_right_key = clock();
        } else {
            window.draw(up);
        }
    } else if (key_state == 3) {
        if ((clock() - std::max(timer_left_key, timer_right_key)) / CLOCKS_PER_SEC > BONGO_KEYPRESS_THRESHOLD) {
            window.draw(wave);
            timer_wave_key = clock();
        } else {
            window.draw(up);
        }
    }

    // draw smoke
    bool is_smoke_key_pressed = false;

    for (Json::Value &v : smoke_key_value) {
        if (input::is_pressed(v.asInt())) {
            is_smoke_key_pressed = true;
            break;
        }
    }

    if (is_enable_toggle_smoke) {
        previous_smoke_key_state = current_smoke_key_state;
        current_smoke_key_state = is_smoke_key_pressed;

        bool is_smoke_key_down = current_smoke_key_state && (current_smoke_key_state != previous_smoke_key_state);

        if (is_smoke_key_down) {
            is_toggle_smoke = !is_toggle_smoke;
        }
    }
    else {
        is_toggle_smoke = is_smoke_key_pressed;
    }

    if (is_toggle_smoke) {
        window.draw(smoke);
    }
}
}; // namespace osu


src/main.cpp
#include "header.hpp"

#if !defined(__unix__) && !defined(__unix)
#include <windows.h>
#endif

sf::RenderWindow window;

#if defined(__unix__) || defined(__unix)
int main(int argc, char ** argv) {
#else
int APIENTRY WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow) {
#endif
    if (AttachConsole(ATTACH_PARENT_PROCESS)) {
        freopen("CONOUT$","wb",stdout);
        freopen("CONOUT$","wb",stderr);
    }

    // loading configs
    while (!data::init()) {
        continue;
    }
    window.create(sf::VideoMode(data::cfg["width"].asInt(), data::cfg["height"].asInt()), "Bongocat v2.0.0", sf::Style::Titlebar | sf::Style::Close);
    window.setFramerateLimit(MAX_FRAMERATE);

    // initialize input
    if (!input::init()) {
        return EXIT_FAILURE;
    }

    bool is_reload = false;
    bool is_show_input_debug = false;

    while (window.isOpen()) {
        sf::Event event;
        while (window.pollEvent(event)) {
            switch (event.type) {
            case sf::Event::Closed:
                window.close();
                break;

            case sf::Event::KeyPressed:
                // get reload config prompt
                if (event.key.code == sf::Keyboard::R && event.key.control) {
                    if (!is_reload) {
                        while (!data::init()) {
                            continue;
                        }
                    }
                    is_reload = true;
                    break;
                }

                // toggle joystick debug panel
                if (event.key.code == sf::Keyboard::D && event.key.control) {
                    // is_show_input_debug = !is_show_input_debug;
                    osu::toggle_debug();
                    break;
                }

            default:
                is_reload = false;
            }
        }

        int mode = data::cfg["mode"].asInt();

        Json::Value rgb = data::cfg["decoration"]["rgb"];
        int red_value = rgb[0].asInt();
        int green_value = rgb[1].asInt();
        int blue_value = rgb[2].asInt();
        int alpha_value = rgb.size() == 3 ? 255 : rgb[3].asInt();

        window.clear(sf::Color(red_value, green_value, blue_value, alpha_value));
        switch (mode) {
        case 1:
            osu::draw();
            break;
        case 2:
            taiko::draw();
            break;
        case 3:
            ctb::draw();
            break;
        case 4:
            mania::draw();
            break;
        case 5:
            custom::draw();
        }

        if (is_show_input_debug) {
            input::drawDebugPanel();
        }

        window.display();
    }

    input::cleanup();
    return 0;
}



include/header.hpp
#pragma once
#define BONGO_KEYPRESS_THRESHOLD 0
#define MAX_FRAMERATE 60

#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
#include <map>

#include <time.h>

#include <math.h>
#include <string.h>

#include <SFML/Graphics.hpp>
#include "json/json.h"

extern sf::RenderWindow window;

namespace data {
extern Json::Value cfg;
extern std::string skin;

void error_msg(std::string error, std::string title);

bool init();

sf::Texture &load_texture(std::string path);
}; // namespace data

namespace input {
bool init();

bool is_pressed(int key_code);

bool is_joystick_connected();
bool is_joystick_pressed(int key_code);

std::pair<double, double> bezier(double ratio, std::vector<double> &points, int length);

std::pair<double, double> where_mouse();
std::pair<double, double> get_xy();

void drawDebugPanel();

void cleanup();
}; // namespace input

namespace osu {
void toggle_debug();

bool init();

void draw();
}; // namespace osu

namespace taiko {
bool init();

void draw();
}; // namespace taiko

namespace ctb {
bool init();

void draw();
}; // namespace ctb

namespace mania {
bool init();

void draw();
}; // namespace mania

namespace custom {
bool init();

void draw();
}; // namespace custom


src/data.cpp
#include "header.hpp"
#define BONGO_ERROR 1

#if defined(__unix__) || defined(__unix)
#include <unistd.h>
#include <limits.h>

extern "C" {
#include <SDL2/SDL.h>
}
#else
#include <windows.h>
#endif

namespace data {
Json::Value cfg;
std::string skin = "default";
std::map<std::string, sf::Texture> img_holder;

void create_config() {
    const char *s =
        R"V0G0N({
    "mode": 1,
    "width": 612,
    "height": 352,
    "resolution": {
        "letterboxing": false,
        "width": 1920,
        "height": 1080,
        "horizontalPosition": 0,
        "verticalPosition": 0
    },
    "decoration": {
        "leftHanded": false,
        "rgb": [255, 255, 255],
        "offsetX": [0, 11],
        "offsetY": [0, -65],
        "scalar": [1.0, 1.0]
    },
    "mouse": true,
    "toggleSmoke": false,
    "paw": [255, 255, 255],
    "pawEdge": [0, 0, 0],
    "key1": [90],
    "key2": [88],
    "smoke": [67],
    "wave": [],
    "stretchyArm": true,
    "debug": true,
    "anchor": [162, 110],
    "armAnchorOffset": [67, 0],
    "catsMouse": [60, 176, 180, 100, 15]
    "taiko": {
        "leftCentre": [88],
        "rightCentre": [67],
        "leftRim": [90],
        "rightRim": [86]
    },
    "catch": {
        "left": [37],
        "right": [39],
        "dash": [16]
    },
    "mania": {
        "4K": true,
        "key4K": [68, 70, 74, 75],
        "key7K": [83, 68, 70, 32, 74, 75, 76]
    },
    "custom": {
        "background": "img/osu/mousebg.png",
        "mouse": false,
        "mouseImage": "img/osu/mouse.png",
        "mouseOnTop": true,
        "offsetX": 0,
        "offsetY": 0,
        "scalar": 1.0,
        "paw": [255, 255, 255],
        "pawEdge": [0, 0, 0],
        "keyContainers": []
    },
    "mousePaw": {
        "mousePawComment": "coordinates start in the top left of the window",
        "pawStartingPoint": [211, 159],
        "pawEndingPoint": [258, 228]
    }
})V0G0N";
    std::string error;
    Json::CharReaderBuilder cfg_builder;
    Json::CharReader *cfg_reader = cfg_builder.newCharReader();
    cfg_reader->parse(s, s + strlen(s), &cfg, &error);
    delete cfg_reader;
}

void error_msg(std::string error, std::string title) {
#if defined(__unix__) || defined(__unix)

    SDL_MessageBoxButtonData buttons[] = {
        { SDL_MESSAGEBOX_BUTTON_RETURNKEY_DEFAULT, 0, "Retry" },
        { SDL_MESSAGEBOX_BUTTON_ESCAPEKEY_DEFAULT, 1, "Cancel" },
    };

    SDL_MessageBoxColorScheme colorScheme = {
        { /* .colors (.r, .g, .b) */
     /* [SDL_MESSAGEBOX_COLOR_BACKGROUND] */
        { 255, 255,255 },
        /* [SDL_MESSAGEBOX_COLOR_TEXT] */
        { 0, 0, 0 },
        /* [SDL_MESSAGEBOX_COLOR_BUTTON_BORDER] */
        { 0, 0, 0 },
        /* [SDL_MESSAGEBOX_COLOR_BUTTON_BACKGROUND] */
        { 255,255, 255 },
        /* [SDL_MESSAGEBOX_COLOR_BUTTON_SELECTED] */
        { 128, 128, 128 }
        }
    };

    SDL_MessageBoxData messagebox_data = {
    	SDL_MESSAGEBOX_ERROR,
    	NULL,
    	title.c_str(),
    	error.c_str(),
    	SDL_arraysize(buttons),
    	buttons,
    	&colorScheme
    };

    int button_id;

    SDL_ShowMessageBox(&messagebox_data, &button_id);

    if (button_id == -1 || button_id == 1) {
#else
    if (MessageBoxA(NULL, error.c_str(), title.c_str(), MB_ICONERROR | MB_RETRYCANCEL) == IDCANCEL) {
#endif
        exit(BONGO_ERROR);
    }
}

bool update(Json::Value &cfg_default, Json::Value &cfg) {
    bool is_update = true;
    for (const auto &key : cfg.getMemberNames()) {
        if (cfg_default.isMember(key)) {
            if (cfg_default[key].type() != cfg[key].type()) {
                error_msg("Value type error in config.json", "Error reading configs");
                return false;
            }
            if (cfg_default[key].isArray() && !cfg_default[key].empty()) {
                for (Json::Value &v : cfg[key]) {
                    if (v.type() != cfg_default[key][0].type()) {
                        error_msg("Value type in array error in config.json", "Error reading configs");
                        return false;
                    }
                }
            }
            if (cfg_default[key].isObject()) {
                is_update &= update(cfg_default[key], cfg[key]);
            } else {
                cfg_default[key] = cfg[key];
            }
        } else {
            cfg_default[key] = cfg[key];
        }
    }
    return is_update;
}

std::string config_path() {
    std::ifstream skin_list("skin.txt");
    if (skin_list.good()) {
        getline(skin_list, skin);
    }
    std::cout << skin << std::endl;
    return "skins/" + skin + "/config.json";
}

bool init() {
    while (true) {
        std::string path = config_path();
        create_config();
        std::ifstream cfg_file(path, std::ifstream::binary);
        if (!cfg_file.good()) {
            break;
        }
        std::string cfg_string((std::istreambuf_iterator<char>(cfg_file)), std::istreambuf_iterator<char>()), error;
        Json::CharReaderBuilder cfg_builder;
        Json::CharReader *cfg_reader = cfg_builder.newCharReader();
        Json::Value cfg_read;
        if (!cfg_reader->parse(cfg_string.c_str(), cfg_string.c_str() + cfg_string.size(), &cfg_read, &error)) {
            delete cfg_reader;
            error_msg("Syntax error in config.json:\n" + error, "Error reading configs");
        } else if (update(cfg, cfg_read)) {
            delete cfg_reader;
            break;
        }
    }

    img_holder.clear();

    int mode = data::cfg["mode"].asInt();

    switch (mode) {
    case 1:
        return osu::init();
    case 2:
        return taiko::init();
    case 3:
        return ctb::init();
    case 4:
        return mania::init();
    case 5:
        return custom::init();
    default:
        error_msg("Mode value is not correct", "Error reading configs");
        return false;
    }
}

sf::Texture &load_texture(std::string path) {
    if (img_holder.find(path) == img_holder.end()) {
        while (!img_holder[path].loadFromFile(path)) {
            error_msg("Cannot find file " + path, "Error importing images");
        }
    }
    return img_holder[path];
}
}; // namespace data
